8349241: Fix the concurrent execution JVM crash of StringBuilder::append(int/long)

wenshao opened the pull request:
The following code can reproduce the problem, writing out of bounds causes JVM Crash

```java
         StringBuilder buf = new StringBuilder();
        buf.append('中');

        Thread[] threads = new Thread[40];
        final CountDownLatch latch = new CountDownLatch(threads.length);
        Runnable r = () -> {
            for (int i = 0; i < 1000000; i++) {
                buf.setLength(0);
                buf.trimToSize();
                buf.append(123456789123456789L);
            }
            latch.countDown();
        };

        for (int i = 0; i < threads.length; i++) {
            threads[i] = new Thread(r);
        }
        for (Thread t : threads) {
            t.start();
        }
        latch.await();
```

This problem can be avoided by using the value of ensureCapacityInternal directly.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8349241](https://bugs.openjdk.org/browse/JDK-8349241): Fix the concurrent execution JVM crash of StringBuilder::append(int/long) (**Bug** - P3)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23427/head:pull/23427` \
`$ git checkout pull/23427`

Update a local copy of the PR: \
`$ git checkout pull/23427` \
`$ git pull https://git.openjdk.org/jdk.git pull/23427/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23427`

View PR using the GUI difftool: \
`$ git pr show -t 23427`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23427.diff">https://git.openjdk.org/jdk/pull/23427.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23427#issuecomment-2631817922)
</details>

liach commented on src/java.base/share/classes/java/lang/AbstractStringBuilder.java:
```diff
@@ -838,7 +841,7 @@ public AbstractStringBuilder append(char c) {
         int spaceNeeded = count + DecimalDigits.stringSize(i);
-        ensureCapacityInternal(spaceNeeded);
+        byte[] value = ensureCapacityInternal(spaceNeeded);
         if (isLatin1()) {
```
This is not safe. The ensureCapacityInternal can read coder == LATIN1 and allocate a small array, but this `isLatin1` can read coder == UTF16 and write a UTF16 number out of bounds.

cl4es replied:
A check that `spaceNeeded < (value.length >> 1)`  in the `else` branch would be needed and might be a sufficient safeguard here.

wenshao replied:
I made further improvements to improve the thread safety of the coder by passing the newCapacity method into the coder. I think this should be safe enough.

AlanBateman commented:
/reviewers 2 reviewer

tstuefe commented:
I am confused about the description. How exactly is the JVM crashed? Does the interpreter or compiled code crash?

wenshao commented:
> ```java
>  StringBuilder buf = new StringBuilder();
>     buf.append('中');
> 
>     final CountDownLatch latch = new CountDownLatch(10);
>     Runnable r = () -> {
>         for (int i = 0; i < 10000; i++) {
>             buf.setLength(0);
>             buf.trimToSize();
>             buf.append(123456789);
>         }
>         latch.countDown();
>     };
>     Thread[] threads = new Thread[10];
>     for (int i = 0; i < threads.length; i++) {
>         threads[i] = new Thread(r);
>     }
>     for (Thread t : threads) {
>         t.start();
>     }
>     latch.await();
> ```

Will cause the JVM to exit directly, the error message is as follows
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0xa) at pc=0x0000000103b9d23c, pid=23348, tid=33539
#
# JRE version: OpenJDK Runtime Environment (25.0) (build 25-internal-adhoc.wenshao.jdkx)
# Java VM: OpenJDK 64-Bit Server VM (25-internal-adhoc.wenshao.jdkx, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# [thread 41731 also had an error]
V  [libjvm.dylib+0x41523c]  G1ParScanThreadState::copy_to_survivor_space(G1HeapRegionAttr, oopDesc*, markWord)+0x64
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
[thread 39683 also had an error]
# An error report file with more information is saved as:
# /Users/wenshao/Work/git/jdk_mico_bench/hs_err_pid23348.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
```

tstuefe commented:
@wenshao Thank you. This seems to be a GC problem. I adjusted the JBS issue accordingly. You set this to "24" as affected version, but if this is a mainline issue, please add 25 and if possible all other versions this occurs in. If possible, please attach an hs-err file or at least the crash stack.

wenshao commented:
> @wenshao Thank you. This seems to be a GC problem. I adjusted the JBS issue accordingly. You set this to "24" as affected version, but if this is a mainline issue, please add 25 and if possible all other versions this occurs in. If possible, please attach an hs-err file or at least the crash stack.


I added the hs-err file in the reply above. This is not a GC problem. The getChars method uses StringUTF16.putChar, which is equivalent to Unsafe.putChar. There is no out-of-bounds check. When concurrent, out-of-bounds writes will occur, causing JVM Crash.

liach commented:
On a second examination, I find this is caused by #22023, which is not in 24. And I cannot replicate the oob write on 24. I have updated the affected version to 25 as a result, and updated the caused-by link in the JBS issue.

RogerRiggs commented on src/java.base/share/classes/java/lang/AbstractStringBuilder.java:
```diff
@@ -260,7 +275,7 @@ private void ensureCapacityInternal(int minimumCapacity) {
      *         greater than (Integer.MAX_VALUE >> coder)
      */
-    private int newCapacity(int minCapacity) {
+    private static int newCapacity(int minCapacity, byte[] value, byte coder) {
```
1.  Only the current length is needed and should be the argument.
2. Add the \@param tags for the new arguments.

RogerRiggs commented on src/java.base/share/classes/java/lang/AbstractStringBuilder.java:
```diff
@@ -241,11 +241,26 @@ public void ensureCapacity(int minimumCapacity) {
+     */
+    private byte[] ensureCapacityInternal(int minimumCapacity, byte coder) {
+        // overflow-conscious code
+        byte[] value = this.value;
```
Shadowing `value` is a bug prone, pick a new name.

RogerRiggs commented on src/java.base/share/classes/java/lang/AbstractStringBuilder.java:
```diff
@@ -241,11 +241,26 @@ public void ensureCapacity(int minimumCapacity) {
+     * synchronized.
+     * If {@code minimumCapacity} is non positive due to numeric
+     * overflow, this method throws {@code OutOfMemoryError}.
+     */
```
Complete the javadoc with the \@param tags and descriptions.

wenshao commented:
Thanks @RogerRiggs, your suggestion is great, I have fixed it, please help me review it again.

tstuefe commented:
> > @wenshao Thank you. This seems to be a GC problem. I adjusted the JBS issue accordingly. You set this to "24" as affected version, but if this is a mainline issue, please add 25 and if possible all other versions this occurs in. If possible, please attach an hs-err file or at least the crash stack.
> 
> I added the hs-err file in the reply above. This is not a GC problem. The getChars method uses StringUTF16.putChar, which is equivalent to Unsafe.putChar. There is no out-of-bounds check. When concurrent, out-of-bounds writes will occur, causing JVM Crash.

@wenshao I see. Yes, you are right. Interesting - I was not aware of JDK code using unsafe-like put calls internally.

wenshao commented:
keep alive
